<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>GenoPop</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">GenoPop</h1>



<div id="welcome-to-the-genopop-manual" class="section level1">
<h1>Welcome to the GenoPop-Manual</h1>
<p>GenoPop is a R package designed to assist with population genomic
analyses of data sets from non-model organisms or with low sequencing
quality. It’s created with the intention to simplify and streamline the
analysis of large genomic data sets in VCF (Variant Call Format) files
in a efficient manner, while handling problems of missing data.</p>
<p>The GenoPop can be divided into parts. One part it the genotype
imputation method GenoPop-impute, and the second part contains several
function to calculate commonly used population genomics metrics, like
Fst, and Dxy.</p>
<p>This document will give an overview about GenoPops functions and
usability. Starting with an overview of its methods, a guide how to
install and get started with the package.</p>
<div id="genopop-impute-overview" class="section level2">
<h2>GenoPop-Impute Overview</h2>
<p>GenoPop-Impute is a genotype imputation algorithm specifically
designed for whole genome data sets. It does not require a SNP reference
panel and, therefore, can be used for non-model organisms. A key aspect
of GenoPop-Impute is its approach to handling large genomic datasets.
Recognizing that SNPs within a linkage block share the same evolutionary
history, GenoPop-Impute employs the assumption that these SNPs exhibit
more comparable patterns than those from different linkage blocks. This
assumption justifies segmenting the dataset into smaller blocks for
parallel processing. Essentially, GenoPop-Impute performs batch-based
imputation, where each batch contains SNPs likely to be correlated due
to their close proximity in the genome and linkage disequilibrium. Using
the missForest algorithm (Stekhoven &amp; Bühlmanm, 2012) for the
imputation of each batch, this approach enhances the efficiency and
accuracy of the imputation process. A guide on how to execute
GenoPop-Impute can be found using the R help options.</p>
</div>
<div id="population-genomics-metrics-overview" class="section level2">
<h2>Population Genomics Metrics Overview</h2>
<div id="general-functionality" class="section level3">
<h3>General Functionality</h3>
<p>Each function in this part of GenoPop is designed to calculate
specific population genomics metrics directly from VCF (Variant Call
Format) files. These functions are designed for efficiency and handle
large genomic datasets by processing data in parallel. For this, there
are two modi available: processing in batches of equal numbers of SNPs
and processing in windows of a specific genomic size in base pairs. In
batch mode, the entire VCF file is processed at once to provide a
general overview. In window mode, the file is processed in genomic
sections to identify specific regions of interest. These functions
typically return single metrics for batch mode or data frames detailing
metrics per window.</p>
</div>
<div id="metrics-overview" class="section level3">
<h3>Metrics Overview</h3>
<ul>
<li><p><strong>FixedSites</strong>: Counts the number of sites fixed for
the alternative allele. It helps identify regions with a complete
fixation of an allele, potentially indicating selective sweeps or other
evolutionary pressures.</p></li>
<li><p><strong>SegregatingSites</strong>: Counts the number of
polymorphic sites, which are not fixed for the alternative allele. It’s
a measure of genetic variability within the population.</p></li>
<li><p><strong>SingletonSites</strong>: Counts the number of singleton
sites, where a minor allele occurs only once in the sample. It can be an
indicator of recent mutations.</p></li>
<li><p><strong>PrivateAlleles</strong>: Calculates the number of private
alleles in two populations. Private alleles are present in one
population but absent in another, providing insight into population
differentiation.</p></li>
<li><p><strong>ObservedHeterozygosity (Ho)</strong>: Calculates the
observed heterozygosity for each variant. It’s a measure of genetic
diversity within a population.</p></li>
<li><p><strong>ExpectedHeterozygosity (He)</strong>: Calculates the
expected heterozygosity based on allele frequencies. It’s a theoretical
measure of how genetically diverse a population should be under random
mating.</p></li>
<li><p><strong>NucleotideDiversity (Pi)</strong>: Measures the average
number of nucleotide differences per site between two sequences. It’s a
key indicator of genetic diversity within a population.</p></li>
<li><p><strong>Tajima’s D</strong>: A neutrality test comparing the
number of segregating sites to the average number of nucleotide
differences. It can suggest population expansion, selection, or
bottlenecks.</p></li>
<li><p><strong>Watterson’s Theta</strong>: A measure of genetic
diversity within a population, based on the number of segregating
sites.</p></li>
<li><p><strong>Average Nucleotide Differences (Dxy)</strong>: Measures
the average number of nucleotide differences per site between two
populations. It’s a measure of genetic differentiation.</p></li>
<li><p><strong>Fst</strong>: The fixation index, measuring genetic
differentiation between populations. It ranges from 0 (no
differentiation) to 1 (complete differentiation).</p></li>
<li><p><strong>OneDimSFS</strong>: Calculates a one-dimensional site
frequency spectrum, either folded or unfolded. It provides insights into
allele frequency distributions within a population.</p></li>
<li><p><strong>TwoDimSFS</strong>: Calculates a two-dimensional site
frequency spectrum for two populations. It’s used to infer demographic
history and population relationships.</p></li>
</ul>
<p>Please note that this summary provides an overview of the functions
and their purposes. For complete understanding and appropriate usage,
refer to the detailed documentation of each function.</p>
</div>
</div>
<div id="installation-instructions-for-genopop" class="section level2">
<h2>Installation Instructions for GenoPop</h2>
<div id="prerequisites" class="section level3">
<h3>Prerequisites</h3>
<p>Before installing GenoPop, make sure you have R installed on your
system. You can download and install R from <a href="https://cran.r-project.org/">CRAN</a>.</p>
</div>
<div id="installing-genopop-from-github" class="section level3">
<h3>Installing GenoPop from GitHub</h3>
<p>To install the GenoPop package directly from GitHub, you will need
the <code>devtools</code> package in R. If you don’t have
<code>devtools</code> installed, you can install it by running the
following command in R:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span></code></pre></div>
<p>Once devtools is installed, you can install GenoPop using the
install_github function. Run the following commands in your R
console:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">&quot;https://github.com/MiGurke/GenoPop&quot;</span>)</span></code></pre></div>
<p>Then load the package..</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(GenoPop)</span></code></pre></div>
<p>and your ready to go!</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
